qiskit/optimization/converters/optimization_problem_to_negative_value_oracle.py

```diff
@@ -0,0 +1,230 @@
+        """
+        Constructor.
+        Args:
+            key_bits (int): The number of key bits.
+            value_bits (int): The number of value bits.
+            ancilla_bits (int): The number of precision (result) bits.
+            func_dict (dict): The dictionary of function coefficients to encode.
+            backend (str, optional): A string corresponding to the name of the selected backend.
+        """
+        QuantumDictionary.__init__(self, key_bits, value_bits, ancilla_bits, func_dict,
+                                   self.prepare_quadratic, backend=backend)
+        self._circuit = None
+
+    @property
+    def circuit(self):
+        """
+        Provides the circuit of the Quantum Dictionary. Will construct one if not yet created.
+        Returns:
+            QuantumCircuit: Circuit object describing the Quantum Dictionary.
+        """
+        if self._circuit is None:
+            self._circuit = self.construct_circuit()
+        return self._circuit
+
+    def prepare_quadratic(self, func_dict, circuit, key_val):
+        """
+        Encodes a QUBO in the proper dictionary format into the state of a given register.
+        Args:
+            func_dict (dict): The QUBO problem. The keys should be subscripts of the coefficients
+                (e.g. x_1 -> 1), with the constant (if present) being represented with a key of -1
+                (i.e. d[-1] = constant). Quadratic coefficients should use a tuple for the key, with
+                 the corresponding subscripts inside (e.g. 2*x_1*x_2 -> d[(1,2)]=2).
+            circuit (QuantumCircuit): The circuit to apply the operator to.
+            key_val (QuantumRegister): Register containing the key and value qubits. They are
+                combined here to follow the register format expected by algorithms like Qiskit
+                Aqua's Grover.
+        """
+        circuit.h(key_val)
+
+        # Linear Coefficients
+        for i in range(self.value_bits):
+            if func_dict.get(-1, 0) != 0:
+                circuit.u1(1 / 2 ** self.value_bits * 2 * np.pi * 2 ** i * func_dict[-1],
+                           key_val[self.key_bits + i])
+            for j in range(self.key_bits):
+                if func_dict.get(j, 0) != 0:
+                    circuit.cu1(1 / 2 ** self.value_bits * 2 * np.pi * 2 ** i * func_dict[j],
+                                key_val[j], key_val[self.key_bits + i])
+
+        # Quadratic Coefficients
+        for i in range(self.value_bits):
+            for k, v in func_dict.items():
+                if isinstance(k, tuple):
+                    circuit.mcu1(1/2 ** self.value_bits * 2 * np.pi * 2 ** i * v,
+                                 [key_val[k[0]], key_val[k[1]]], key_val[self.key_bits + i])
+
+        iqft = IQFT(self.value_bits)
+        value = [key_val[v] for v in range(self.key_bits, self.key_bits + self.value_bits)]
+        iqft.construct_circuit(qubits=value, circuit=circuit)
```

Comment:
See comment in the general description: 
> Introducing the `QuantumDictionary` in this format seems like a lot of overhead for implementing the functionality of L209-L230. Also, the naming suggests to be a fundamental object similar to the `QuantumRegister` or `QuantumCircuit`.  
Since this PR is about the Grover Optimization Algorithm and contains already a lot of changes: would it be possible to move the functionality of the `QQUBODictionary` (i.e. L209-L230) directly in the converter? 

> If the `QuantumDictionary` is something we should add in Qiskit, this should go into Terra and be discussed there. But as that includes some overhead I think it would be better to not include it in this PR.

Reply:
This might be a question for @stefan-woerner. 

In the past we discussed adding it as a class because it may be converted to a utility in the future. It's pretty handy for encoding key-value pairs - we have other extensions of the QDictionary that encode generic functions (not as efficient as QUBO), calculates total sums, etc.). If that's no longer the case, or not in scope at the moment, I can boil down the functionality and just include it directly in the converter.

Reply:
Moved in the upcoming commit.